File: objectivec/day17_part1_2021.m
#import <Foundation/Foundation.h>

BOOL isMovingAway(int xPos, int yPos, int xVel, int yVel, int xMin, int xMax, int yMin, int yMax) {
    return (xPos < xMin && xVel < 0) || (xPos > xMax && xVel > 0) || (yPos < yMin && yVel < 0);
}

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        NSString *input = [NSString stringWithContentsOfFile:@"input.txt" encoding:NSUTF8StringEncoding error:nil];
        NSArray *parts = [input componentsSeparatedByString:@", "];
        NSArray *xRange = [[parts[0] substringFromIndex:15] componentsSeparatedByString:@".."];
        NSArray *yRange = [[parts[1] substringFromIndex:2] componentsSeparatedByString:@".."];
        
        int xMin = [xRange[0] intValue], xMax = [xRange[1] intValue];
        int yMin = [yRange[0] intValue], yMax = [yRange[1] intValue];
        
        int maxY = INT_MIN;
        for (int xVel = -1000; xVel <= 1000; xVel++) {
            for (int yVel = -1000; yVel <= 1000; yVel++) {
                int xPos = 0, yPos = 0, curXVel = xVel, curYVel = yVel, highestY = yPos;
                while (1) {
                    xPos += curXVel;
                    yPos += curYVel;
                    if (xPos >= xMin && xPos <= xMax && yPos >= yMin && yPos <= yMax) {
                        if (highestY > maxY) maxY = highestY;
                        break;
                    }
                    if (isMovingAway(xPos, yPos, curXVel, curYVel, xMin, xMax, yMin, yMax)) break;
                    if (curXVel > 0) curXVel--;
                    else if (curXVel < 0) curXVel++;
                    curYVel--;
                    if (yPos > highestY) highestY = yPos;
                }
            }
        }
        NSLog(@"%d", maxY);
    }
    return 0;
}

File: elixir/day17_part1_2021.exs

defmodule Solution do
  def solve do
    {:ok, input} = File.read("input.txt")
    [x_range, y_range] = String.split(input, ", ")
    
    [x_min, x_max] = 
      x_range 
      |> String.slice(15..-1) 
      |> String.split("..")
      |> Enum.map(&String.to_integer/1)
    
    [y_min, y_max] = 
      y_range 
      |> String.slice(2..-1) 
      |> String.split("..")
      |> Enum.map(&String.to_integer/1)
    
    result = 
      for xVel <- -1000..1000,
          yVel <- -1000..1000,
          reduce: -1 do
        max_y -> 
          case simulate(xVel, yVel, x_min, x_max, y_min, y_max) do
            {:hit, height} -> max(max_y, height)
            :miss -> max_y
          end
      end
    
    IO.puts(result)
  end
  
  defp simulate(x_vel, y_vel, x_min, x_max, y_min, y_max) do
    do_simulate(0, 0, x_vel, y_vel, x_min, x_max, y_min, y_max, 0)
  end
  
  defp do_simulate(x_pos, y_pos, x_vel, y_vel, x_min, x_max, y_min, y_max, highest_y) 
       when x_pos >= x_min and x_pos <= x_max and y_pos >= y_min and y_pos <= y_max do
    {:hit, highest_y}
  end
  
  defp do_simulate(x_pos, y_pos, x_vel, y_vel, x_min, x_max, y_min, y_max, highest_y) 
       when x_pos < x_min and x_vel < 0 or 
            x_pos > x_max and x_vel > 0 or 
            y_pos < y_min and y_vel < 0 do
    :miss
  end
  
  defp do_simulate(x_pos, y_pos, x_vel, y_vel, x_min, x_max, y_min, y_max, highest_y) do
    new_x_pos = x_pos + x_vel
    new_y_pos = y_pos + y_vel
    
    new_x_vel = 
      cond do
        x_vel > 0 -> x_vel - 1
        x_vel < 0 -> x_vel + 1
        true -> 0
      end
    
    new_y_vel = y_vel - 1
    new_highest_y = max(highest_y, new_y_pos)
    
    do_simulate(
      new_x_pos, 
      new_y_pos, 
      new_x_vel, 
      new_y_vel, 
      x_min, 
      x_max, 
      y_min, 
      y_max, 
      new_highest_y
    )
  end
end

Solution.solve()
